src/lib.rs

```diff
@@ -3,83 +3,80 @@
 
 //! rust-vmm device model.
 
-use std::cmp::{Ord, Ordering, PartialOrd};
-
+pub mod bus;
 pub mod device_manager;
 pub mod resources;
 
-// IO Size.
-#[derive(Debug, Copy, Clone)]
-enum IoSize {
-    // Port I/O size.
-    Pio(u16),
+use std::ops::Deref;
+use std::sync::{Arc, Mutex};
+
+use bus::{MmioAddress, PioAddress, PioAddressValue};
 
-    // Memory mapped I/O size.
-    Mmio(u64),
+pub trait DevicePio {
+    fn pio_read(&self, base: PioAddress, offset: PioAddressValue, data: &mut [u8]);
+    fn pio_write(&self, base: PioAddress, offset: PioAddressValue, data: &[u8]);
```

Comment:
I think we should make PIO and MMIO ops return a `Result`. There could be several ways for an IO ops to fail and we should let the VMM decide what to do with such failure. For example, when the guest tries to read or write from/to an unregistered device, this is silenced by us not returning an error. Some VMMs may want to ignore that situation, others may decide it's a security problem and act upoin it.

Reply:
Hi Sam! Regarding you particular example, the device ops from `DevicePio` and `DeviceMmio` only get invoked after the bus has validated there is a device registered with the range in question. We can definitely have errors as part of executing the operation, but previous discussions on this topic concluded that returning a result is probably not very useful since the vCPU thread that ends up handling the VM exit & IO access is not necessarily in a better position to handle errors compared to the device itself. 

When more elaborated error handling is required, the proposed approach was to communicate with different parts of the application via some backchannel. Representing the error type is also quite awkward (unless it's a predefined type for all potential devices); we'd mostly likely have to add a  generic type parameter to the traits. Do you think not returning a result here is a blocker going forward, or we can proceed with the current approach as well?

Reply:
Hey @alexandruag If we're ok discussing this trait through further PRs and potentially break the API (We're far from having a release for that crate, so I guess it should be fine...), it's not a blocker. I do remember the discussions we had on that topic.

I think there are many ways to report generic errors carrying device specific payload and allow the VMM for at least reporting device errors and potentially acting upon it. Things like unregistered ranges, ranges crossing multiple devices, ranges overflowing expected address spaces, and generic IO error carrying e.g.  an `anyhow::Error` device specific payload would already be useful.

Reply:
Yeah, I think it's totally reasonable to keep this conversation open and bring changes (even breaking ones) in subsequent PRs, especially since we haven't even published a `0.X.Y` version yet (AFAIK, things are considered more stable starting with `1.0.0`). We can also create an issue to track the discussion if you think that's helpful.